### ISSUE TITLE
[feat] New rule: No IIFEs in JSX

### ISSUE THREAD

**Issue by kachkaev** [Rel1cx/eslint-react#1112]
### Describe the problem

I came across a component that looked like this:

```tsx
function SomeComponent() {
  // hooks etc

  return (
    <SomeJsx>
      <SomeMoreJsx />

      {(() => {
        const filteredThings = things.filter(callback);

        if (filteredThings.length === 0) {
          return <Empty />;
        }

        return filteredThings.map((thing) => <Thing key={thing.id} data={thing} />);
      })()}

      <SomeMoreJsx />
    </SomeJsx>
  );
}
```

So there is an IIFE (Immediately Invoked Function Expression) inside JSX, which is used to define some additional variables and logic.

Technically, this is valid JS, but it is not conventional inside React components. IIFEs in JSX may be hard to follow and they will probably not optimized by React compiler, which means slower app rendering.

### Describe the solution you'd like

I'd like to have a rule that detects IIFEs in JSX and flags them. If we need to define additional variables, we should do it before JSX. If the logic is too heavy and we want to keep it separate, we can create a new React component.

### Alternatives considered

It may be possible to ban IIFEs via [no-restricted-syntax](https://eslint.org/docs/latest/rules/no-restricted-syntax) but that’s harder to configure and maintain.

**Comment by laug** [comment]
I'm not sure if this is the right place to have this discussion, but is there any evidence that IIFEs **_will probably not optimized by React compiler_** as is written above? After some research, I was not able to find any. Of course depending on the contents of the IIFEs, if those do things like violate the rules of React that would be a problem, but that does not mean there is a problem with IIFEs themselves.

Also, you could replace any IIFE:
```js
    <SomeJsx>
      <SomeMoreJsx />

      {(() => {
        const filteredThings = things.filter(callback);

        if (filteredThings.length === 0) {
          return <Empty />;
        }

        return filteredThings.map((thing) => <Thing key={thing.id} data={thing} />);
      })()}

      <SomeMoreJsx />
    </SomeJsx>
```
with the below code:
```js
    <SomeJsx>
      <SomeMoreJsx />

      {[0].map(x => {
        const filteredThings = things.filter(callback);

        if (filteredThings.length === 0) {
          return <Empty />;
        }

        return filteredThings.map((thing) => <Thing key={thing.id} data={thing} />);
      })}

      <SomeMoreJsx />
    </SomeJsx>
```
which is exactly equivalent, but does not get flagged, and is supposedly ok to use with React Compiler as it is not an IIFE. Everybody agrees that using `.map()` is fine, but I don't see why the compiler would have trouble with the IIFE version if it is able to handle the `map` version.

Regarding IIFEs being "hard to follow", that is a subjective observation, but most of the time the purpose of the IIFE is just to define a local variable that needs to be reused more than once in the logic. In that case, the simplest fix to remove the IIFE is moving that variable definition up just before the `return` statement, but that arguably makes the code _harder_ to follow than the IIFE version given the variable declaration and the related code that uses it are now completely separate.
